firebase-sessions/src/main/kotlin/com/google/firebase/sessions/leader/SessionMaintainerLeader.kt

```diff
@@ -0,0 +1,170 @@
+import com.google.firebase.sessions.SessionGenerator
+import com.google.firebase.sessions.SessionInitiateListener
+import com.google.firebase.sessions.SessionInitiator
+import com.google.firebase.sessions.SessionMaintainer
+import com.google.firebase.sessions.Time
+import com.google.firebase.sessions.TimeProvider
+import com.google.firebase.sessions.api.FirebaseSessionsDependencies
+import com.google.firebase.sessions.api.SessionSubscriber
+import com.google.firebase.sessions.settings.SessionsSettings
+import kotlinx.coroutines.CoroutineDispatcher
+
+/**
+ * Maintainer that will manage the full lifecycle of a session including sending this to our
+ * backend. It will also notify followers when there is a change
+ */
+class SessionMaintainerLeader(
+  private val firebaseApp: FirebaseApp,
+  firebaseInstallations: FirebaseInstallationsApi,
+  backgroundDispatcher: CoroutineDispatcher,
+  blockingDispatcher: CoroutineDispatcher,
+  transportFactoryProvider: Provider<TransportFactory>,
+) : SessionMaintainer {
+
+  private val applicationInfo = SessionEvents.getApplicationInfo(firebaseApp)
+  private val sessionSettings =
+    SessionsSettings(
+      firebaseApp.applicationContext,
+      blockingDispatcher,
+      backgroundDispatcher,
+      firebaseInstallations,
+      applicationInfo,
+    )
+  private val timeProvider: TimeProvider = Time()
+  private val sessionGenerator =
+    SessionGenerator(collectEvents = shouldCollectEvents(), timeProvider)
+  private val eventGDTLogger = EventGDTLogger(transportFactoryProvider)
+  private val sessionCoordinator = SessionCoordinator(firebaseInstallations, eventGDTLogger)
+
+  private val TAG = "SessionMaintainerLeader"
+
+  override fun start(backgroundDispatcher: CoroutineDispatcher) {
+    val sessionInitiateListener =
+      object : SessionInitiateListener {
+        // Avoid making a public function in FirebaseSessions for onInitiateSession.
+        override suspend fun onInitiateSession(sessionDetails: SessionDetails) {
+          initiateSessionStart(sessionDetails)
+        }
+      }
+
+    val sessionInitiator =
+      SessionInitiator(
+        timeProvider,
+        backgroundDispatcher,
+        sessionInitiateListener,
+        sessionSettings,
+        sessionGenerator,
+      )
+
+    val appContext = firebaseApp.applicationContext.applicationContext
+    if (appContext is Application) {
+      appContext.registerActivityLifecycleCallbacks(sessionInitiator.activityLifecycleCallbacks)
+
+      firebaseApp.addLifecycleEventListener { _, _ ->
+        Log.w(TAG, "FirebaseApp instance deleted. Sessions library will not collect session data.")
+        appContext.unregisterActivityLifecycleCallbacks(sessionInitiator.activityLifecycleCallbacks)
+      }
+    } else {
+      Log.e(
+        TAG,
+        "Failed to register lifecycle callbacks, unexpected context ${appContext.javaClass}."
+      )
+    }
+  }
+  override fun register(subscriber: SessionSubscriber) {
```

Comment:
Will register need to be different for leaders and followers? I thought this would stay in the FirebaseSessions class.

Reply:
Yes. Followers will pass something different. The will pass different things.